src/mam4xx/modal_aer_opt.hpp

```diff
@@ -0,0 +1,1084 @@
+
+        // get aerosol properties and save for each species
+        // Fortran to C++ indexing
+        auto specmmr = state_q_kk[lmassptr_amode[ll][mm] - 1];
+        // Fortran to C++ indexing
+        const Real specdens = specdens_amode[lspectype_amode[ll][mm] - 1];
+
+        // allocate(specvol(pcols,nspec),stat=istat)
+        specvol[ll] = specmmr / specdens;
+      } // ll
+
+      // lw =0 and sw =1
+      Real dryvol, wetvol, watervol = {};
+      Kokkos::complex<Real> crefin = {};
+      Real refr, refi = {};
+
+      calc_refin_complex(1, isw, qaerwat_m_kk[mm], specvol,
+                         aersol_optics_data.specrefindex_sw[mm], nspec,
+                         aersol_optics_data.crefwlw, aersol_optics_data.crefwsw,
+                         dryvol, wetvol, watervol, crefin, refr, refi);
+
+      // interpolate coefficients linear in refractive index
+      // first call calcs itab,jtab,ttab,utab
+
+      const auto sub_extpsw = aersol_optics_data.extpsw[mm][isw];
+      const auto ref_real_tab = aersol_optics_data.refrtabsw[mm][isw];
+      const auto ref_img_tab = aersol_optics_data.refitabsw[mm][isw];
+
+      int itab = zero;   // index for Bilinear interpolation
+      int itab_1 = zero; // ! index for Bilinear interpolation column 1
+      // FIXME: part of binterp only apply for first column
+      int jtab = zero;
+      Real ttab, utab = {}; // coef for Bilinear interpolation
+      Real cext[ncoef], cabs[ncoef],
+          casm[ncoef] = {}; //  ! coefficient for extinction, absoption, and
+                            //  asymmetry [unitless]
+
+      binterp(sub_extpsw, refr, refi, ref_real_tab.data(), ref_img_tab.data(),
+              itab, jtab, ttab, utab, cext, itab_1);
+
+      const auto sub_abspsw = aersol_optics_data.abspsw[mm][isw];
+
+      binterp(sub_abspsw, refr, refi, ref_real_tab.data(), ref_img_tab.data(),
+              itab, jtab, ttab, utab, cabs, itab_1);
+
+      const auto sub_asmpsw = aersol_optics_data.asmpsw[mm][isw];
+
+      binterp(sub_asmpsw, refr, refi, ref_real_tab.data(), ref_img_tab.data(),
+              itab, jtab, ttab, utab, casm, itab_1);
+
+      // parameterized optical properties
+      Real pext = zero; //    parameterized specific extinction [m2/kg]
+      calc_parameterized(cext, cheb_kk, pext);
+      Real pabs = zero; // parameterized specific absorption [m2/kg]
+      calc_parameterized(cabs, cheb_kk, pabs);
+
+      Real pasm = zero; // parameterized asymmetry factor [unitless?]
+      calc_parameterized(casm, cheb_kk, pasm);
+
+      //  do icol=1,ncol
+
+      if (logradsurf <= xrmax) {
+        pext = haero::exp(pext);
+      } else {
+        // BAD CONSTANT
+        pext = 1.5 / (radsurf * rhoh2o); // ! geometric optics
+      }                                  // if logradsurf(kk) <= xrmax
+
+      // convert from m2/kg water to m2/kg aerosol
+      // FIXME: specpext is used by check_error_warning, which is not ported
```

Comment:
Are we going to port `check_error_warning`, or do we not need to worry about this?

Reply:
According to the Fortran them the condition inside these function is never met on testing run. Thus, I will not port this code.